File: 107.BinaryTreeLevelOrderTraversal2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def levelOrderBottom(self, root: Optional[TreeNode]) -> List[List[int]]:
        levels = []
        queue = []

        if root == None:
            return levels
        
        queue.append(root)
        queue.append(None)

        level = []

        while len(queue) > 0:
            ele = queue.pop(0)

            if ele == None:
                levels.append(level)
                level = []
                if len(queue) == 0:
                    break
                else:
                    queue.append(None)
            else:
                level.append(ele.val)
                if ele.left != None:
                    queue.append(ele.left)
                if ele.right != None:
                    queue.append(ele.right)

        return levels[::-1]


        

File: 107.BinaryTreeLevelOrderTraversal2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {

        List<List<Integer>> levels = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();

        List<Integer> level = new ArrayList<>();

        if (root == null) {
            return levels;
        }

        queue.add(root);
        queue.add(null);

        while(!queue.isEmpty()) {
            TreeNode ele = queue.remove();
            if (ele == null) {
                levels.add(level);
                level = new ArrayList<>();
                if(queue.isEmpty()) {
                    break;
                } else {
                    queue.add(null);
                }
            } else {
                level.add(ele.val);

                if (ele.left != null) {
                    queue.add(ele.left);
                }
                if (ele.right != null) {
                    queue.add(ele.right);
                }
            }
        }

        List<List<Integer>> res = new ArrayList<>();

        for (int i = levels.size() - 1; i >=0; i--) {
            res.add(levels.get(i));
        }

        return res;
        
    }
}
